File: src/java/com/springbok/pattern/PatternERR_011V01.java
/* Copyright (C) 2022 Springbok LLC

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
package com.springbok.pattern;

import com.springbok.utility.PatternUtility;

import java.util.Map;

/**
 * Describes the ITU antenna pattern ERR_011V01.
 */
public class PatternERR_011V01 implements EarthPattern, TransmitPattern {

	/** Maximum antenna gain [dB] */
	private double GainMax;

	/** Diameter of an earth antenna [m] */
	private double Diameter;

	/* Intermediate gain calculation results */
	/* None */

	/**
	 * Constructs a PatternERR_011V01 given a maximum antenna gain.
	 *
	 * @param GainMax Maximum antenna gain [dB]
	 * @param Diameter Diameter of an earth antenna [m]
	 */
	public PatternERR_011V01(double GainMax, double Diameter) {

		/* Validate input parameters */
		PatternUtility.validate_input("GainMax", GainMax);
		PatternUtility.validate_input("Diameter", Diameter);

		/* Assign properties */
		this.GainMax = GainMax;
		this.Diameter = Diameter;

		/* Compute derived properties */
		/* None */
	}

	/** 
	 * Gets maximum antenna gain [dB].
	 * 
	 * @return Maximum antenna gain [dB]
	 */
	public double getGainMax() {
		return GainMax;
	}

	/**
	 * Gets diameter of an earth antenna [m].
	 * 
	 * @return Diameter of an earth antenna [m]
	 */
	public double getDiameter() {
		return Diameter;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see com.springbok.pattern.EarthPattern#copy()
	 */
	public PatternERR_011V01 copy() {
		return new PatternERR_011V01(this.GainMax, this.Diameter);
	}

	/**
	 * Appendix 30A (RR-2003) reference transmitting earth station antenna pattern
	 * for Region 2.
	 *
	 * @param Phi Angle for which a gain is calculated [degrees]
	 * @param options Optional parameters (entered as key/value pairs)
	 * 
	 * @return Co-polar and cross-polar gain [dB
	 */
	public Gain gain(double Phi, Map options) {

		/* Validate input parameters */
		PatternUtility.validate_input("Phi", Phi);

		/* Validate input options */
		options = PatternUtility.validateOptions(options);

		/* Implement pattern */
		double eps = Math.ulp(1.0);
		double phi = Math.max(eps, Phi);
		double G0 = GainMax;
		double G1 = 36.0 - 20.0 * Math.log10(phi);
		double G2 = 51.3 - 53.2 * Math.pow(phi, 2);
		double G3 = Math.max(29.0 - 25.0 * Math.log10(phi), -10.0);

		double G = G0 * (0 <= phi && phi < 0.1 ? 1 : 0)
				+ G1 * (0.1 <= phi && phi < 0.32 ? 1 : 0)
				+ G2 * (0.32 <= phi && phi < 0.54 ? 1 : 0)
				+ G3 * (0.54 <= phi && phi <= 180 ? 1 : 0);

		G = Math.min(GainMax, G);

		double Gx0 = GainMax - 30;
		double Gx1 = Math.max(9.0 - 20.0 * Math.log10(phi), -10.0);

		double phi_x = 0.6 / Diameter;

		double Gx = Gx0 * (0 <= phi && phi < phi_x ? 1 : 0)
				+ Gx1 * (phi_x <= phi && phi <= 180 ? 1 : 0);

		Gx = Math.min(GainMax - 30, Gx);

		/* Validate low level rules */
		if (G < Gx) {
			throw new IllegalStateException("Co-polar curve is less than cross-polar curve [6013: STDC_ERR_CO_LT_CX]");
		}
		if (Diameter < 2.5) {
			throw new IllegalStateException("Diameter is less that 2.5 [6007: STDC_ERR_Diameter]");
		}

		/* Validate output parameters */
		if ((boolean) options.get("DoValidate")) {
			PatternUtility.validate_output(G, Gx, GainMax);
		}
		return new Gain(G, Gx);
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see java.lang.Object#hashCode()
	 */
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		long temp;
		temp = Double.doubleToLongBits(GainMax);
		result = prime * result + (int) (temp ^ (temp >>> 32));
		temp = Double.doubleToLongBits(Diameter);
		result = prime * result + (int) (temp ^ (temp >>> 32));
		return result;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see java.lang.Object#equals(java.lang.Object)
	 */
	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null) {
			return false;
		}
		if (!(obj instanceof PatternERR_011V01)) {
			return false;
		}
		PatternERR_011V01 other = (PatternERR_011V01) obj;
		if (Double.doubleToLongBits(GainMax) != Double.doubleToLongBits(other.GainMax)) {
			return false;
		}
		if (Double.doubleToLongBits(Diameter) != Double.doubleToLongBits(other.Diameter)) {
			return false;
		}
		return true;
	}
}

File: src/matlab/com/springbok/pattern/PatternERR_011V01.m
% Copyright (C) 2022 Springbok LLC
% 
% This program is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, either version 3 of the License, or (at
% your option) any later version.
% 
% This program is distributed in the hope that it will be useful, but
% WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
% General Public License for more details.
% 
% You should have received a copy of the GNU General Public License
% along with this program.  If not, see <https://www.gnu.org/licenses/>.
% 
classdef PatternERR_011V01 < EarthPattern & TransmitPattern
% Describes the ITU antenna pattern ERR_011V01.
  
  properties (SetAccess = private, GetAccess = public)
    
    % Maximum antenna gain [dB]
    GainMax
    % Diameter of an earth antenna, m
    Diameter
    
    % Intermediate gain calculation results
    
  end % properties (SetAccess = private, GetAccess = public)
  
  methods
    
    function this = PatternERR_011V01(GainMax, Diameter)
    % Constructs a PatternERR_011V01 given a maximum antenna gain.
    %
    % Parameters
    %   GainMax - Maximum antenna gain [dB]
    %   Diameter - Diameter of an earth antenna, m
      
    % No argument constructor
      if nargin == 0
        return;
        
      end % if
      
      % Check number and class of input arguments.
      if ~(nargin == 2 ...
           && isnumeric(GainMax) ...
           && isnumeric(Diameter))
        exc = SException( ...
            'Springbok:InvalidInput', ...
            'Numeric GainMax, and Diameter must be provided.');
        throw(exc)
        
      end % if
      
      % Validate input parameters.
      PatternUtility.validate_input('GainMax', GainMax);
      PatternUtility.validate_input('Diameter', Diameter);
      
      % Assign properties
      this.GainMax = GainMax;
      this.Diameter = Diameter;
      
      % Compute derived properties
      % None
      
    end % PatternERR_011V01()
    
    function that = copy(this)
    % Copies a PatternERR_011V01 given a maximum antenna gain.
    %
    % Parameters
    %   None
      that = PatternERR_011V01(this.GainMax, this.Diameter);

    end % copy()

    function [G, Gx, fH] = gain(this, Phi, varargin)
    % Appendix 30A (RR-2003) reference transmitting earth station
    % antenna pattern for Region 2.
    %
    % Parameters:
    %   Phi - Angle for which a gain is calculated [degrees]
    %
    % Optional parameters (entered as key/value pairs):
    %   PlotFlag - Flag to indicate plot, true or {false}
    %
    % Returns:
    %   G - Co-polar gain [dB]
    %   Gx - Cross-polar gain [dB]
    %   fH - figure handle
      
    % Check number and class of input arguments.
      if ~(nargin >= 1 ...
           && isnumeric(Phi))
        exc = SException( ...
            'Springbok:InvalidInput', ...
            'Numeric Phi must be provided.');
        throw(exc)
        
      end % if
      
      % Validate input parameters.
      PatternUtility.validate_input('Phi', Phi);
      
      % Handle optional inputs.
      input = struct();
      if ~isempty(varargin)
        nVarargin = size(varargin, 2);
        
        if rem(nVarargin, 2)
          exc = SException( ...
              'Springbok:InvalidInput', ...
              'Inputs must be key/value pairs.');
          throw(exc)
          
        else
          for iVarargin = 1:2:nVarargin
            key = varargin{iVarargin};
            value = varargin{iVarargin + 1};
            input = PatternUtility.validate_input(key, value, input);
            
          end % for
          
        end % if
        
      end % if
      
      % Assign default values to any optional input parameters that
      % were not included in varargin.
      if ~isfield(input, 'PlotFlag')
        PlotFlag = false;

      else
        PlotFlag = input.PlotFlag;

      end % if
      if ~isfield(input, 'DoValidate')
        DoValidate = true;

      else
        DoValidate = input.DoValidate;

      end % if

      % Implement pattern.
      phi = max(eps, Phi); 
      
      G0 = this.GainMax;
      G1 = 36 - 20 * log10(phi);
      G2 = 51.3 - 53.2 * phi .^ 2;
      G3 = max(29 - 25 * log10(phi), -10);
      
      G = G0 .* (0    <= phi & phi <   0.1) ...
          + G1 .* (0.1  <= phi & phi <  0.32) ...
          + G2 .* (0.32 <= phi & phi <  0.54) ...
          + G3 .* (0.54 <= phi & phi <=  180);
      
      G = min(this.GainMax, G);
      
      Gx0 = this.GainMax - 30;
      Gx1 = max(9 - 20 * log10(phi), -10);
      
      phi_x = 0.6 / this.Diameter;
      
      Gx = Gx0 .* (0     <= phi & phi <  phi_x) ...
           + Gx1 .* (phi_x <= phi & phi <=   180);
      
      Gx = min(this.GainMax - 30, Gx);
      
      % Validate low level rules.
      if G < Gx
        exc = SException( ...
            'Springbok:InvalidResult', ...
            'Co-polar curve is less than cross-polar curve.', ...
            -6013, ...
            'STDC_ERR_CO_LT_CX', ...
            'Co-polar curve is less than cross-polar curve.');
        throw(exc)
        
      end % if
      
      if this.Diameter < 2.5
        exc = SException( ...
            'Springbok:InvalidResult', ...
            'Diameter is less that 2.5.', ...
            -6007, ...
            'STDC_ERR_Diameter', ...
            'Diameter is less that 2.5.');
        throw(exc)
        
      end % if
      
      % Validate output parameters.
      if DoValidate
        PatternUtility.validate_output(G, Gx, this.GainMax);
      
      end % if
      
      % Plot results, if requested.
      if PlotFlag == true
        fH = figure;
        semilogx(phi, G, 'k')
        grid on
        hold on
        semilogx(phi, Gx, 'k')
        v = axis;
        axis([0.01 180 v(3) v(4)])
        title([strrep(mfilename, '_', '\_'), ' Reference Antenna Pattern'])
        xlabel('Phi (degrees)')
        ylabel('Gain (dB)')
        
      else
        fH = [];
        
      end % if
      
    end % gain()
    
  end % methods
  
end % classdef
